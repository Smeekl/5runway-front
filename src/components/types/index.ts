import { InputProps } from 'antd';

export type TSimpleWrapper = {
  className?: string;
};

export type TSearchInput = {
  className?: string;
  startIcon?: boolean;
  endIcon?: boolean;
  onIconClick?: () => void;
} & InputProps;

export type TSidebarErrorIndicator = {
  content: string;
  active: boolean;
};

export type TSidebarCustomFormikField = {
  name: string;
  className: string;
};

export type TSearchIcon = {
  onIconClick?: () => void;
};

export type TCustomFormikSlider = {
  minCurrentValue: number;
  maxCurrentValue: number;
  minValue: number;
  maxValue: number;
  minInputName: string;
  maxInputName: string;
  step: number;
  format: string;
  setFieldValue: (field: string, value: unknown) => void;
};
