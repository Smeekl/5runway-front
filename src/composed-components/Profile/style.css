.profile {
  padding-top: 65px;
  width: 100%;
}

.profile__wrapper {
  padding: 0 35px;
}

.profile__title {
  font-weight: 700;
  font-size: 24px;
  color: var(--color-primary);
}

.profile__email {
  font-weight: 300;
  font-size: 12px;
}

.profile__subscribe {
  font-weight: 300;
  font-size: 12px;
  margin-bottom: 18px;
}

.profile__name {
  margin-right: 3px;
}

.profile__subscribe-type {
  color: var(--color-primary-blue);
}

.profile__subscribe-type:hover {
  opacity: 0.7;
}

.settings__link {
  padding: 16px 35px;
  display: inline-block;
  font-weight: 300;
  font-size: 12px;
  color: var(--color-primary);
  width: 100%;
  height: 100%;
}

.settings__link > span > svg > path {
  transition: 0.2s all;
}

.settings__link:hover > span > svg > path {
  fill: var(--color-primary-blue);
}

.settings__logout {
  transition: 0.2s all;
}

.settings__link:hover .settings__logout {
  color: var(--color-primary-blue);
}

.settings__item {
  border-bottom: 1px solid var(--color-primary-light);
}

.settings__item:first-child {
  border-top: 1px solid var(--color-primary-light);
}

.settings__link_icon {
  display: flex;
  justify-content: space-between;
  align-items: center;
}
