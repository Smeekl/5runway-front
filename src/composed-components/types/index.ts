import React from 'react';

export type TSearchProvider = {
  children: React.ReactElement;
};

export type TSearchValue = {
  history: string[];
  clearHistory: () => void;
  setInHistory: () => void;
  setInputValue: (value: string) => void;
  inputValue: string;
  handleSetHistory: () => void;
};

export type TUser = {
  data: {
    email: string;
    subscriptionPlan: string;
    expiringDate: number;
  };
};
