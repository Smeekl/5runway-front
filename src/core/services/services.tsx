import React from 'react';
import { IServices } from '@serviceTypes';

const { createContext, useContext } = React;
const ServicesContext = createContext({} as IServices);

type TServiceProviderProps = {
  services: IServices;
  children: React.ReactNode;
};

export const ServiceProvider = ({
  services,
  children,
}: TServiceProviderProps) => (
  <ServicesContext.Provider value={services}>
    {children}
  </ServicesContext.Provider>
);

export const useService = () => useContext(ServicesContext);
