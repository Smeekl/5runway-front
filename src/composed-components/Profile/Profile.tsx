import { Link, useLocation } from 'react-router-dom';
import { ExportOutlined } from '@ant-design/icons';
import { useGetUser } from './useGetUser.hook';
import './style.css';

const Profile = () => {
  const location = useLocation();
  const { isLoading, user } = useGetUser();

  return (
    <div className="profile">
      {isLoading ? (
        'Loading...'
      ) : (
        <div className="profile__wrapper">
          <h3 className="profile__title">Profile</h3>
          <p className="profile__email">{user?.email}</p>
          <p className="profile__subscribe">
            <span className="profile__name">Subscribe:</span>
            <Link to="/" className="profile__subscribe-type">
              <span>{user?.subscriptionPlan || 'Free'}</span>{' '}
              <span>
                {user?.expiringDate! > 0 ? user?.expiringDate : 'Unsubscribe'}
              </span>
            </Link>
          </p>
        </div>
      )}
      <ul className="profile__settings settings">
        <li className="settings__item">
          <Link className="settings__link" to="/">
            YOUR AIRCRAFTS
          </Link>
        </li>
        <li className="settings__item">
          <Link className="settings__link" to="/">
            ADD AIRCRAFT
          </Link>
        </li>
        <li className="settings__item">
          <Link
            className="settings__link settings__link_icon"
            to={`${location.pathname}/logout`}
          >
            <span className="settings__logout">LOG OUT</span>
            <ExportOutlined />
          </Link>
        </li>
      </ul>
    </div>
  );
};

export default Profile;
