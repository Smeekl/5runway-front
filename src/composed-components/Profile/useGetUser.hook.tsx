import { useQuery } from 'react-query';
import { useService } from '@services/services';

export const useGetUser = () => {
  const { userService } = useService();

  const {
    data: userData,
    isLoading,
    isError,
    error,
    refetch,
  } = useQuery('userData', () => userService.getUser());

  return {
    user: userData?.data,
    error,
    isLoading,
    isError,
    refetch,
  };
};
