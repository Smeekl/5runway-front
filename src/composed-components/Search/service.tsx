import React, { useState, useEffect } from 'react';
import { useService } from '@services/services';
import { TSearchProvider, TSearchValue } from '@composed-components/types';

const { createContext, useContext } = React;
const SearchContext = createContext({} as TSearchValue);

export const SearchProvider = ({ children }: TSearchProvider) => {
  const { searchService } = useService();

  const [history, setHistory] = useState<Array<string>>([]);
  const [inputValue, setInputValue] = useState('');

  useEffect(() => {
    setHistory(searchService.getSearchedItems() || []);
  }, []);

  function setInHistory() {
    if (!inputValue.trim()) return;

    searchService.searchItem(inputValue);
    setHistory((prevHistory) => [inputValue, ...prevHistory.slice(0, 9)]);
    setInputValue('');
  }

  function clearHistory() {
    searchService.clearSearchedItems();
    setHistory([]);
  }

  function handleSetHistory() {
    if (history.length <= 10) {
      return;
    }

    setHistory([...history.slice(0, 10)]);
  }

  return (
    <SearchContext.Provider
      value={{
        setInHistory,
        clearHistory,
        history,
        inputValue,
        setInputValue,
        handleSetHistory,
      }}
    >
      {children}
    </SearchContext.Provider>
  );
};

export const useSearch = () => useContext(SearchContext);
