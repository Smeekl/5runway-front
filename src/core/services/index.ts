import { CreateApiClient } from './apiClient';
import { StorageService } from './storage-services/storage.services';
import { UserService } from './user.service';
import { SearchService } from './search.service';
import { AuthService } from './auth.service';

export function createServices() {
  const storageService = new StorageService(window.localStorage);
  const api = CreateApiClient(process.env.REACT_APP_BASEURL, storageService);
  const searchService = new SearchService(api, storageService);
  const userService = new UserService(api, storageService);
  const authService = new AuthService(api, storageService);

  return {
    storageService,
    searchService,
    userService,
    authService,
  };
}
