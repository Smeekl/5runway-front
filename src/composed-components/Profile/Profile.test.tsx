import { QueryClient, QueryClientProvider } from 'react-query';
import { MemoryRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import { waitFor } from '@testing-library/dom';
import { ServiceProvider } from '@services/services';
import { UserService } from '@services/user.service';
import { TUser } from '@composed-components/types/index';
import Profile from './Profile';

describe('Profile UI tests', () => {
  let queryClient: QueryClient = null;

  beforeEach(() => {
    queryClient = new QueryClient({
      defaultOptions: {
        queries: {
          staleTime: Infinity,
        },
      },
    });
  });

  test('should render correctly', () => {
    const { container } = render(
      <MemoryRouter>
        <QueryClientProvider client={queryClient}>
          <Profile />
        </QueryClientProvider>
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  test('should be in the document', () => {
    const { container } = render(
      <MemoryRouter>
        <QueryClientProvider client={queryClient}>
          <Profile />
        </QueryClientProvider>
      </MemoryRouter>
    );

    expect(container).toBeInTheDocument();
  });

  test('should render data after response', async () => {
    const userServiceSpy = jest.spyOn(new UserService({}, {}), 'getUser');
    const delay = (value: TUser): Promise<TUser> =>
      new Promise((resolve) => setTimeout(resolve, 50, value));

    userServiceSpy.mockImplementation(() =>
      delay({
        data: {
          email: 'bigBebra@gmail.com',
          expiringDate: 3,
          subscriptionPlan: 'Premium',
        },
      })
    );

    render(
      <MemoryRouter>
        <ServiceProvider
          services={{ userService: { getUser: userServiceSpy } }}
        >
          <QueryClientProvider client={queryClient}>
            <Profile />
          </QueryClientProvider>
        </ServiceProvider>
      </MemoryRouter>
    );

    expect(screen.getByText('Loading...')).toBeInTheDocument();

    await waitFor(
      () => {
        expect(userServiceSpy).toHaveBeenCalledTimes(1);

        expect(screen.getByText('bigBebra@gmail.com')).toBeInTheDocument();
        expect(screen.getByText('3')).toBeInTheDocument();
        expect(screen.getByText('Premium')).toBeInTheDocument();
      },
      { timeout: 100 }
    );
  });
});
