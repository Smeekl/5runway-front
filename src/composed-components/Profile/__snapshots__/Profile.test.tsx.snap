// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile UI tests should render correctly 1`] = `
<div>
  <div
    class="profile"
  >
    Loading...
    <ul
      class="profile__settings settings"
    >
      <li
        class="settings__item"
      >
        <a
          class="settings__link"
          href="/"
        >
          YOUR AIRCRAFTS
        </a>
      </li>
      <li
        class="settings__item"
      >
        <a
          class="settings__link"
          href="/"
        >
          ADD AIRCRAFT
        </a>
      </li>
      <li
        class="settings__item"
      >
        <a
          class="settings__link settings__link_icon"
          href="//logout"
        >
          <span
            class="settings__logout"
          >
            LOG OUT
          </span>
          <span
            aria-label="export"
            class="anticon anticon-export"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="export"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M888.3 757.4h-53.8c-4.2 0-7.7 3.5-7.7 7.7v61.8H197.1V197.1h629.8v61.8c0 4.2 3.5 7.7 7.7 7.7h53.8c4.2 0 7.7-3.4 7.7-7.7V158.7c0-17-13.7-30.7-30.7-30.7H158.7c-17 0-30.7 13.7-30.7 30.7v706.6c0 17 13.7 30.7 30.7 30.7h706.6c17 0 30.7-13.7 30.7-30.7V765.1c0-4.3-3.5-7.7-7.7-7.7zm18.6-251.7L765 393.7c-5.3-4.2-13-.4-13 6.3v76H438c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h314v76c0 6.7 7.8 10.5 13 6.3l141.9-112a8 8 0 000-12.6z"
              />
            </svg>
          </span>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
